README.md
# avtool

[![License: MIT](https://img.shields.io/badge/License-GPL_v3-brightgreen.svg)](https://github.com/clok/avtool/blob/main/LICENSE)
[![Go Report Card](https://goreportcard.com/badge/clok/avtool)](https://goreportcard.com/report/clok/avtool)
[![Coverage Status](https://coveralls.io/repos/github/clok/avtool/badge.svg?branch=main)](https://coveralls.io/github/clok/avtool?branch=main)
[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/clok/avtool/v2?tab=overview)

> NOTE: Original code written by [@pbthorste](https://github.com/pbthorste) for [https://github.com/pbthorste/avtool](https://github.com/pbthorste/avtool)
>
> HUGE SHOUT OUT to [@pbthorste](https://github.com/pbthorste)

This module provides a reimplementation of `ansible-vault` encrypt and decprypt functionality in Go.

## Why the fork?

As of writing the mainline has been stale for ~4 years.

I have found this code to be highly useful and important for writing other `ansible-vault` related tools. I wanted to modernize the work done previously to support `go.mod` while also updating the interface as an importable module for other code.

## CLI Tool

Please see [gwvault](https://github.com/GoodwayGroup/gwvault) for a purpose built `ansible-vault` binary written in go.

It leverages the work done by [@pbthorste](https://github.com/pbthorste) for [https://github.com/pbthorste/avtool](https://github.com/pbthorste/avtool) while further fleshing out the CLI tool to be more in line with the original `ansible-vault` CLI tool.

## Thanks and Attribution

Original code written by [@pbthorste](https://github.com/pbthorste)


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v2.1.1"></a>
## [v2.1.1] - 2021-05-13
### Bug Fixes
- **deps:** update golang.org/x/crypto commit hash to cd7d49e ([#7](https://github.com/clok/avtool/issues/7))

### Chore
- **ci:** added Makefile and goreleaser


<a name="v2.1.0"></a>
## [v2.1.0] - 2021-03-15
### Chore
- **ci:** port to using golangci-lint github action
- **deps:** update actions/setup-go action to v2 ([#4](https://github.com/clok/avtool/issues/4))
- **deps:** update actions/checkout action to v2 ([#3](https://github.com/clok/avtool/issues/3))
- **go.mod:** bump to go 1.16

### Features
- **release:** v2.1.0


<a name="v2.0.3"></a>
## [v2.0.3] - 2021-03-03
### Chore
- **README:** update badges

### Features
- **release:** v2.0.3


<a name="v2.0.2"></a>
## [v2.0.2] - 2021-03-03
### Bug Fixes
- **github actions:** point to forked go-proxy-pull-action to support major versions

### Features
- **release:** v2.0.2


<a name="v2.0.1"></a>
## [v2.0.1] - 2021-03-03
### Chore
- **github actions:** update github actions versions
- **github actions:** add go proxy warming

### Features
- **release:** v2.0.1


<a name="v2.0.0"></a>
## [v2.0.0] - 2021-03-02
### Chore
- update README.md
- add CONTRIBUTING.md
- **changelog:** added CHANGELOG.md generated by git-chglog
- **cli:** remove deprecated CLI tool and update README.md
- **lint:** clean up linting issues
- **lint:** clean up code
- **lint:** update for go 1.15 linting rules
- **renovate:** update renovate config
- **tests:** updated tests and added github actions

### Features
- **gomod:** add go.mod support
- **release:** v2.0.0
- **v2:** bump module version to v2


<a name="1.2.0"></a>
## [1.2.0] - 2017-05-25
### Added
- output flag to support vanilla (default) & aligned outputs;
- features to view specific keys/all keys & key names

### Minor
- changes to .gitignore, makefile

### Update
- readme

### Pull Requests
- Merge pull request [#4](https://github.com/clok/avtool/issues/4) from sundarsrd/master


###### Squashed Commits:
```
Added -k flag for selective viewing & additional validations
```



<a name="1.1.1"></a>
## [1.1.1] - 2017-04-21
### Update
- key


<a name="1.1.0"></a>
## [1.1.0] - 2017-04-21
### Add
- travis ci file
- encryption

### Added
- makefile

### Support
- windows line endings

### Use
- version from git


<a name="1.0.0"></a>
## 1.0.0 - 2017-01-24
### Bug Fixes
- usage
- retrievePassword function

### Change
- to GPL v3

### Fm
- t

### Handle
- panic

### Initial
- commit

### Initial
- commit

### Move
- to own function


[Unreleased]: https://github.com/clok/avtool/compare/v2.1.1...HEAD
[v2.1.1]: https://github.com/clok/avtool/compare/v2.1.0...v2.1.1
[v2.1.0]: https://github.com/clok/avtool/compare/v2.0.3...v2.1.0
[v2.0.3]: https://github.com/clok/avtool/compare/v2.0.2...v2.0.3
[v2.0.2]: https://github.com/clok/avtool/compare/v2.0.1...v2.0.2
[v2.0.1]: https://github.com/clok/avtool/compare/v2.0.0...v2.0.1
[v2.0.0]: https://github.com/clok/avtool/compare/1.2.0...v2.0.0
[1.2.0]: https://github.com/clok/avtool/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/clok/avtool/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/clok/avtool/compare/1.0.0...1.1.0


decrypt_test.go
package avtool

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_Decrypt(t *testing.T) {
	password := "asdf"
	content := `$ANSIBLE_VAULT;1.1;AES256
39663038636438383965366163636163376531336238346239623934393436393938656439643133
3638363066366433666438623138373866393763373265320a366635386630336562633763323236
61616562393964666464653532636436346535616566613434613361303734373734383930323661
6664306264366235630a643235323438646132656337613434396338396335396439346336613062
3766
`
	result, err := Decrypt(content, password)
	assert.Equal(t, "hello", result)
	assert.NoError(t, err)
}

func Test_DecryptFile(t *testing.T) {
	password := "asdf"
	filename := "./testdata/test1/secrets.yaml"
	_, err := DecryptFile(filename, password)
	assert.NoError(t, err)
}


encrypt_test.go
package avtool

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_Encrypt(t *testing.T) {
	password := "asdf"
	body := "secret"
	var encrypted string
	var err error
	encrypted, err = Encrypt(body, password)
	assert.NoError(t, err)

	var result string
	result, err = Decrypt(encrypted, password)
	assert.NoError(t, err)
	assert.Equal(t, body, result)
}


decrypt.go
package avtool

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"golang.org/x/crypto/pbkdf2"
	"io/ioutil"
	"log"
	"strings"
)

func check(e error) {
	if e != nil {
		log.Fatalf("error: %v", e)
		panic(e)
	}
}

func DecryptFile(filename, password string) (result string, err error) {
	data, err := ioutil.ReadFile(filename)
	check(err)
	result, err = Decrypt(string(data), password)
	return
}

// Decrypt a string containing the ansible vault
func Decrypt(data, password string) (result string, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("ERROR: %v", r)
		}
	}()
	data = replaceCarriageReturn(data)
	body := splitHeader([]byte(data))
	salt, cryptedHmac, ciphertext := decodeData(body)
	key1, key2, iv := genKeyInitctr(password, salt)
	checkDigest(key2, cryptedHmac, ciphertext)
	aesCipher, err := aes.NewCipher(key1)
	check(err)
	aesBlock := cipher.NewCTR(aesCipher, iv)
	plaintext := make([]byte, len(ciphertext))
	aesBlock.XORKeyStream(plaintext, ciphertext)
	padding := int(plaintext[len(plaintext)-1])
	result = string(plaintext[:len(plaintext)-padding])
	return
}

// in order to support vault files with windows line endings
func replaceCarriageReturn(data string) string {
	return strings.ReplaceAll(data, "\r", "")
}

/*
See _split_header function
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L288
*/
func splitHeader(data []byte) string {
	contents := string(data)
	lines := strings.Split(contents, "\n")
	header := strings.Split(lines[0], ";")
	cipherName := strings.TrimSpace(header[2])
	if cipherName != "AES256" {
		panic("unsupported cipher: " + cipherName)
	}
	body := strings.Join(lines[1:], "")
	return body
}

/*
See decrypt function (in class VaultAES256)
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L741
*/
func decodeData(body string) (salt, cryptedHmac, ciphertext []byte) {
	decoded, _ := hex.DecodeString(body)
	elements := strings.SplitN(string(decoded), "\n", 3)
	salt, err1 := hex.DecodeString(elements[0])
	if err1 != nil {
		panic(err1)
	}
	cryptedHmac, err2 := hex.DecodeString(elements[1])
	if err2 != nil {
		panic(err2)
	}
	ciphertext, err3 := hex.DecodeString(elements[2])
	if err3 != nil {
		panic(err3)
	}
	return
}

/*
See function _gen_key_initctr (in class VaultAES256)
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L685
*/
func genKeyInitctr(password string, salt []byte) (key1, key2, iv []byte) {
	keylength := 32
	ivlength := 16
	key := pbkdf2.Key([]byte(password), salt, 10000, 2*keylength+ivlength, sha256.New)
	key1 = key[:keylength]
	key2 = key[keylength:(keylength * 2)]
	iv = key[(keylength * 2) : (keylength*2)+ivlength]
	return
}

/*
See decrypt function (in class VaultAES256)
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L741
*/
func checkDigest(key2, cryptedHmac, ciphertext []byte) {
	hmacDecrypt := hmac.New(sha256.New, key2)
	_, err := hmacDecrypt.Write(ciphertext)
	check(err)
	expectedMAC := hmacDecrypt.Sum(nil)
	if !hmac.Equal(cryptedHmac, expectedMAC) {
		panic("digests do not match - exiting")
	}
}


encrypt.go
package avtool

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"io/ioutil"
	"strings"
)

func GenerateRandomBytes(n int) ([]byte, error) {
	b := make([]byte, n)
	_, err := rand.Read(b)
	// Note that err == nil only if we read len(b) bytes.
	if err != nil {
		return nil, err
	}
	return b, nil
}

func EncryptFile(filename, password string) (result string, err error) {
	data, err := ioutil.ReadFile(filename)
	check(err)
	result, err = Encrypt(string(data), password)
	return
}

// see https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L710
func Encrypt(body, password string) (result string, err error) {
	salt, err := GenerateRandomBytes(32)
	check(err)
	// salt_64 := "2262970e2309d5da757af6c473b0ed3034209cc0d48a3cc3d648c0b174c22fde"
	// salt,_ = hex.DecodeString(salt_64)
	key1, key2, iv := genKeyInitctr(password, salt)
	ciphertext := createCipherText(body, key1, iv)
	combined := combineParts(ciphertext, key2, salt)
	vaultText := hex.EncodeToString([]byte(combined))
	result = formatOutput(vaultText)
	return
}

func createCipherText(body string, key1, iv []byte) []byte {
	bs := aes.BlockSize
	padding := (bs - len(body)%bs)
	if padding == 0 {
		padding = bs
	}
	padChar := rune(padding)
	padArray := make([]byte, padding)
	for i := range padArray {
		padArray[i] = byte(padChar)
	}
	plaintext := []byte(body)
	plaintext = append(plaintext, padArray...)

	aesCipher, err := aes.NewCipher(key1)
	check(err)
	ciphertext := make([]byte, len(plaintext))

	aesBlock := cipher.NewCTR(aesCipher, iv)
	aesBlock.XORKeyStream(ciphertext, plaintext)
	return ciphertext
}

func combineParts(ciphertext, key2, salt []byte) string {
	hmacEncrypt := hmac.New(sha256.New, key2)
	_, err := hmacEncrypt.Write(ciphertext)
	check(err)
	hexSalt := hex.EncodeToString(salt)
	hexHmac := hmacEncrypt.Sum(nil)
	hexCipher := hex.EncodeToString(ciphertext)
	// nolint:unconvert
	combined := string(hexSalt) + "\n" + hex.EncodeToString([]byte(hexHmac)) + "\n" + string(hexCipher)
	return combined
}

// https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L268
func formatOutput(vaultText string) string {
	heading := "$ANSIBLE_VAULT"
	version := "1.1"
	cipherName := "AES256"

	headerElements := make([]string, 3)
	headerElements[0] = heading
	headerElements[1] = version
	headerElements[2] = cipherName
	header := strings.Join(headerElements, ";")

	elements := make([]string, 1)
	elements[0] = header
	for i := 0; i < len(vaultText); i += 80 {
		end := i + 80
		if end > len(vaultText) {
			end = len(vaultText)
		}
		elements = append(elements, vaultText[i:end])
	}
	elements = append(elements, "")

	whole := strings.Join(elements, "\n")
	return whole
}


cmd/avtool/main.go
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"strings"
	"syscall"

	"github.com/pbthorste/avtool"
	"github.com/smallfish/simpleyaml"
	"golang.org/x/crypto/ssh/terminal"
	cli "gopkg.in/urfave/cli.v1"
)

var (
	version      string
	outputFormat string
)

func main() {
	app := cli.NewApp()
	app.Name = "avtool"
	app.Version = version
	app.Usage = "Tool for working with Ansible Vault files"
	app.Commands = []cli.Command{
		{
			Name:      "view",
			Usage:     "<vaultfile.yml> [-p password] [-k (all|.|<keyname>)] [-f vault-password-file]",
			UsageText: "'<vaultfile.yml>' (view contents of given Ansible Vault file)",
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:  "vault-password-file, f",
					Usage: "load password from 'VAULT_PASSWORD_FILE'",
				},
				cli.StringFlag{
					Name:  "password, p",
					Usage: "`password` to use",
				},
				cli.StringFlag{
					Name:  "key, k",
					Usage: "'key' to retrieve the value for; 'keys' to list just the key names",
					Value: "keys",
				},
				cli.StringFlag{
					Name:  "output, o",
					Usage: "'aligned' to align the output; 'vanilla' for non formatted output",
					Value: "vanilla",
				},
			},
			Action: func(c *cli.Context) error {
				vaultPassword := c.String("vault-password-file")
				password := c.String("password")
				key := c.String("key")
				outputFormat = c.String("output")
				// 01. Input param validations
				err := validateCommandArgs(c)
				if err != nil {
					return err
				}
				vaultFileName, err := validateAndGetVaultFile(c)
				if err != nil {
					return err
				}
				err = validateOutputFormat(c, outputFormat)
				if err != nil {
					return err
				}
				// 02. password retrieval and decryption
				pw, err := retrievePassword(vaultPassword, password)
				if err != nil {
					return cli.NewExitError(err, 2)
				}
				result, err := avtool.DecryptFile(vaultFileName, pw)
				if err != nil {
					return cli.NewExitError(err, 1)
				}
				// 03. print output
				if result != "" {
					secretsYamlBytes := []byte(result)
					secretsYaml, _ := simpleyaml.NewYaml(secretsYamlBytes)
					// parse file contents as yaml
					if secretsYaml != nil {
						// 01. if all names or values are given, retrieve all keys, values
						if key == "all" || key == "keys" {
							keyList, _ := secretsYaml.GetMapKeys()
							println(getDecoratedMessage(fmt.Sprintf("%d Key(s) in %s", len(keyList), vaultFileName)))
							for _, keyName := range keyList {
								println(getDecoratedMessage(keyName))
								if key == "all" {
									println(getYamlKeyValue(keyName, secretsYaml))
								}
							}
						} else {
							// 02. if specific key is given, retrieve only that key
							if outputFormat == "aligned" {
								println(getDecoratedMessage(key))
							}
							println(getYamlKeyValue(key, secretsYaml))
						}
					}
				} else {
					println(getDecoratedMessage(vaultFileName + " is empty!"))
				}
				return nil
			},
		},
		{
			Name:  "encrypt",
			Usage: "vaultfile.yml - encrypt contents of the given yaml file",
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:  "vault-password-file",
					Usage: "load password from `VAULT_PASSWORD_FILE`",
				},
				cli.StringFlag{
					Name:  "password, p",
					Usage: "`password` to use",
				},
				cli.StringFlag{
					Name:  "output, o",
					Usage: "'aligned' to align the output; 'vanilla' for non formatted output",
					Value: "vanilla",
				},
			},
			Action: func(c *cli.Context) error {
				vaultPassword := c.String("vault-password-file")
				password := c.String("password")
				outputFormat = c.String("output")
				// 01. Input param validations
				err := validateCommandArgs(c)
				if err != nil {
					return err
				}
				vaultFileName, err := validateAndGetVaultFile(c)
				if err != nil {
					return err
				}
				err = validateOutputFormat(c, outputFormat)
				if err != nil {
					return err
				}
				//
				pw, err := retrievePassword(vaultPassword, password)
				if err != nil {
					return cli.NewExitError(err, 2)
				}
				result, err := avtool.EncryptFile(vaultFileName, pw)
				if err != nil {
					return cli.NewExitError(err, 1)
				}
				err = ioutil.WriteFile(vaultFileName, []byte(result), 0600)
				if err != nil {
					return cli.NewExitError(err, 1)
				}
				println(getDecoratedMessage("Encryption successful"))
				return nil
			},
		},
	}
	app.Run(os.Args)
}

func validateOutputFormat(c *cli.Context, outputFormat string) (err error) {
	if outputFormat != "vanilla" && outputFormat != "aligned" {
		cli.ShowSubcommandHelp(c)
		return cli.NewExitError(errors.New("ERROR: Unknown output format! Please ref. to usage instructions!"), 2)
	}
	return nil
}

func validateCommandArgs(c *cli.Context) (err error) {
	if !c.Args().Present() {
		cli.ShowSubcommandHelp(c)
		return cli.NewExitError(errors.New(getDecoratedMessage(
			"ERROR: Empty or Invalid inputs! Please ref. to usage instructions!")), 2)
	}
	return nil
}

func validateAndGetVaultFile(c *cli.Context) (filename string, err error) {
	filename = strings.TrimSpace(c.Args().First())
	if filename == "" {
		cli.ShowSubcommandHelp(c)
		return filename, cli.NewExitError(errors.New(getDecoratedMessage(
			"ERROR: Filename not specified!  Please ref. to usage instructions!")), 2)
	} else {
		if fileInfo, err := os.Stat(filename); os.IsNotExist(err) {
			cli.ShowSubcommandHelp(c)
			return filename, cli.NewExitError(errors.New(getDecoratedMessage(
				"ERROR: file "+filename+" "+"doesn't exist!")), 2)
		} else {
			if fileInfo.IsDir() {
				cli.ShowSubcommandHelp(c)
				return filename, cli.NewExitError(errors.New(getDecoratedMessage(
					"ERROR: file "+filename+" is a "+"directory!")), 2)
			}
		}
	}
	// return filename, error on error; nil if no error;
	return filename, nil
}

func getYamlKeyValue(keyName string, secretsYaml *simpleyaml.Yaml) (secretValue string) {
	secretValue, keyErr := secretsYaml.Get(keyName).String()
	if keyErr != nil {
		secretValue = "Key " + keyName + " doesn't exist! " + keyErr.Error()
	}
	return secretValue
}

func retrievePassword(vaultPasswordFile, passwordFlag string) (string, error) {
	if vaultPasswordFile != "" {
		if _, err := os.Stat(vaultPasswordFile); os.IsNotExist(err) {
			return "", errors.New("ERROR: vault-password-file, could not find: " + vaultPasswordFile)
		}
		pw, err := ioutil.ReadFile(vaultPasswordFile)
		if err != nil {
			return "", errors.New("ERROR: vault-password-file, " + err.Error())
		}
		return strings.TrimSpace(string(pw)), nil
	}
	if passwordFlag != "" {
		return passwordFlag, nil
	}

	return readPassword()
}

/*
Reads password from stdin without showing what was entered.
*/
func readPassword() (password string, err error) {
	println(getDecoratedMessage("Enter password: "))
	bytePassword, err := terminal.ReadPassword(int(syscall.Stdin))
	if err != nil {
		err = errors.New("ERROR: could not input password, " + err.Error())
		return
	}
	password = string(bytePassword)
	return
}

func getDecoratedMessage(messageIn string) (messageOut string) {
	messageOut = ""
	// we are using a global variable to control the output format, as getDecoratedMessage is used in multiple places
	// and the format
	if outputFormat == "vanilla" {
		messageOut = messageIn
	} else if outputFormat == "aligned" {
		messageOut = fmt.Sprintf(strings.Repeat(".", 4) + " " + messageIn + " " +
			strings.Repeat(".", 80-len(messageIn)-6))
	}
	return
}
